compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -3404,6 +3404,14 @@ public void visit(BLangMatchStructuredBindingPatternClause patternClause) {
     public void visit(BLangForeach foreach) {
         // Check the collection's type.
         typeChecker.checkExpr(foreach.collection, env);
+        // object type collection should be a subtype of 'object:Iterable
+        if (foreach.collection.type.tag == TypeTags.OBJECT
+                && !types.isAssignable(foreach.collection.type, symTable.iterableType)) {
```

Comment:
Wouldn't this allow in foreach statements iterable objects where the iterator's next may return error too? https://github.com/ballerina-platform/ballerina-lang/issues/28264

Reply:
Is the only place we use iterable is in `foreach`? I changed the error message here please check.
Check https://github.com/ballerina-platform/ballerina-lang/pull/28160/commits/032b8e63532af4fc28843be212b8696d37f0c340